### PR TITLE
Start collecting client-side performance of `Blob` page

### PR DESCRIPTION
**Pull request by olafurpg**
Previously, we didn't have any insight into the user-perceived
performance of loading the blob page. Manual performance tests
indicate that the browser spends a long time calculating HTML/CSS
rules when viewing large files. We have a few ideas on how to optimize
this critical path but first we want to start collecting metrics to
validate that our optimizations are working as expected.

This commit adds a new OpenTelemetry dependency to the React SPA
and registers a new span when loading the blob page. Our backend
consumes the OpenTelemetry spans through the new `/v1/traces` endpoint
and forwards them to an observability tool such as Honeycomb.

The new OpenTelemetry dependencies add ~36kb to the bundle size according to `yarn workspace @sourcegraph/web analyze-bundle`. I think that's quite a lot. We might be able to reduce the overhead by using libhoney directly instead, if we are OK with hardcoding Honeycomb.

CleanShot 2022-06-28 at 16 18 14
CleanShot 2022-06-28 at 16 18 10

## Test plan

n/a

### CODE REVIEW

File: schema/site.schema.json [hunk 1]
```diff
@@ -1095,6 +1095,17 @@
...
       "examples": ["https://sourcegraph.example.com"]
     },
+    "observability.opentelemetry": {
+      "description": "Configuration for OpenTelemetry",
+      "type": "object",
+      "required": ["collectorURL"],
+      "properties": {
+        "collectorURL": {
+          "type": "string",
+          "examples": ["https://COLLECTOR_ENDPOINT:443/v1/traces"]
```
Comment (olafurpg) [inline schema/site.schema.json:1105]: My understanding is that this URL should point to Sourcegraph itself assuming we have added a `/v1/traces` route in frontend. Any suggestions on how to accomplish this?

Comment (Strum355) [reply schema/site.schema.json:1105]: https://github.com/sourcegraph/sourcegraph/blob/olafurpg/client-metrics/schema/site.schema.json#L1093-L1096 seems like an idea, and then add `/v1/traces` or whatever, accessible through https://sourcegraph.com/github.com/sourcegraph/sourcegraph@main/-/blob/client/web/src/jscontext.ts?L34%3A12=

File: client/web/src/repo/blob/BlobPage.tsx [hunk 2]
```diff
@@ -84,10 +86,19 @@ export const BlobPage: React.FunctionComponent<React.PropsWithChildren<Props>> =
         props.location.hash,
     ])
 
+    const tracer = useMemo(() => trace.getTracer('Blob'), [trace])
+
+    const span = useMemo(() => {
+        const span = tracer.startSpan('Blob')
+        span.setAttributes({ repoName, commitID, filePath })
+        return span
+    }, [tracer, repoName, commitID, filePath, renderMode, tracer])
```
Comment (Strum355) [inline client/web/src/repo/blob/BlobPage.tsx:95]: Not sure it makes sense to memoize this (reusing the same span after it has already been ended), but we'd want to differentiate between types of rendering (cold render, render with memoized xyz etc etc)

**Review by valerybugakov — COMMENTED** [review]
Thanks for sharing the bundle size details, @olafurpg! 🙌   Indeed, the difference is quite big. You mentioned [libhoney-js](https://github.com/honeycombio/libhoney-js) in the PR description. Did you have a chance to measure its impact on the bundle size? Curious how it compares to OpenTelemtry SDKs.

I didn't review implementation details since we didn't decide on the SDK we will use for the performance monitoring.
Also, left a comment in [the web app OTel traces issue](https://github.com/sourcegraph/sourcegraph/issues/37907) related to this PR.

**Comment by valerybugakov** [comment]
@olafurpg, upon reading more about `libhoney-js` and other instrumentation libraries available, it seems that all of them are pretty immature compared to backend instrumentation tooling, but `@opentelemetry/*` is the best solution compatible with Honeycomb, and it is OTel compliant which ties nicely into our backend observability strategy. 

Since we won't be changing observability libraries often, we can extract them into a separate Javascript chunk with long-term caching rules (the FPT can help with that) to minimize their impact on loading times. This compromise is worth trying because it will provide much more data for optimization, and with the growing adoption of OTel, benefits will grow over time.

**Comment by olafurpg** [comment]
Thank you @valerybugakov for looking into this. cc/ @oleggromov who may work on telemetry/performance stuff. My gut feeling is that it would be best to extend out existing telemetry service (or a new GQL endpoint) to submit basic performance numbers ("span X took Yms"). I see that  opentelemetry has non-trivial logic related to collecting spans across zones that (I think) we don't have a need for. Something like `const start = Date.now(); ...; const elapsed = Date.now() - start` should be enough

**Comment by olafurpg** [comment]
Closing this PR as dead end for now.

**Pull request closed without merging**